File: moqt/internal/message/fetch.go
package message

import "io"

type FetchMessage struct {
	BroadcastPath string
	TrackName     string
	Priority      uint8
	GroupSequence uint64
}

func (f FetchMessage) Len() int {
	var l int
	l += StringLen(f.BroadcastPath)
	l += StringLen(f.TrackName)
	l += VarintLen(uint64(f.Priority))
	l += VarintLen(f.GroupSequence)
	return l
}

func (f FetchMessage) Encode(w io.Writer) error {
	msgLen := f.Len()
	b := make([]byte, 0, msgLen+VarintLen(uint64(msgLen)))
	b, _ = WriteMessageLength(b, uint64(msgLen))
	b, _ = WriteVarint(b, uint64(len(f.BroadcastPath)))
	b = append(b, f.BroadcastPath...)
	b, _ = WriteVarint(b, uint64(len(f.TrackName)))
	b = append(b, f.TrackName...)
	b, _ = WriteVarint(b, uint64(f.Priority))
	b, _ = WriteVarint(b, f.GroupSequence)
	_, err := w.Write(b)
	return err
}

func (f *FetchMessage) Decode(src io.Reader) error {
	size, err := ReadMessageLength(src)
	if err != nil {
		return err
	}

	b := make([]byte, size)

	_, err = io.ReadFull(src, b)
	if err != nil {
		return err
	}

	str, n, err := ReadString(b)
	if err != nil {
		return err
	}
	f.BroadcastPath = str
	b = b[n:]

	str, n, err = ReadString(b)
	if err != nil {
		return err
	}
	f.TrackName = str
	b = b[n:]

	num, n, err := ReadVarint(b)
	if err != nil {
		return err
	}
	f.Priority = uint8(num)
	b = b[n:]

	num, n, err = ReadVarint(b)
	if err != nil {
		return err
	}
	f.GroupSequence = num
	b = b[n:]

	if len(b) != 0 {
		return ErrMessageTooShort
	}

	return nil
}

File: moq-web/src/internal/message/fetch.ts
import type { Reader, Writer } from "@okdaichi/golikejs/io";
import {
	parseString,
	parseVarint,
	readFull,
	readVarint,
	stringLen,
	varintLen,
	writeString,
	writeVarint,
} from "./message.ts";

export interface FetchMessageInit {
	broadcastPath?: string;
	trackName?: string;
	priority?: number;
	groupSequence?: number;
}

export class FetchMessage {
	broadcastPath: string;
	trackName: string;
	priority: number;
	groupSequence: number;

	constructor(init: FetchMessageInit = {}) {
		this.broadcastPath = init.broadcastPath ?? "";
		this.trackName = init.trackName ?? "";
		this.priority = init.priority ?? 0;
		this.groupSequence = init.groupSequence ?? 0;
	}

	get len(): number {
		return (
			stringLen(this.broadcastPath) +
			stringLen(this.trackName) +
			varintLen(this.priority) +
			varintLen(this.groupSequence)
		);
	}

	async encode(w: Writer): Promise<Error | undefined> {
		const msgLen = this.len;
		let err: Error | undefined;

		[, err] = await writeVarint(w, msgLen);
		if (err) return err;

		[, err] = await writeString(w, this.broadcastPath);
		if (err) return err;

		[, err] = await writeString(w, this.trackName);
		if (err) return err;

		[, err] = await writeVarint(w, this.priority);
		if (err) return err;

		[, err] = await writeVarint(w, this.groupSequence);
		if (err) return err;

		return undefined;
	}

	async decode(r: Reader): Promise<Error | undefined> {
		let err: Error | undefined;

		let msgLen: number;
		[msgLen, , err] = await readVarint(r);
		if (err) return err;

		const buf = new Uint8Array(msgLen);
		[, err] = await readFull(r, buf);
		if (err) return err;

		let offset = 0;

		[this.broadcastPath, offset] = (() => {
			const [str, n] = parseString(buf, offset);
			return [str, offset + n];
		})();

		[this.trackName, offset] = (() => {
			const [str, n] = parseString(buf, offset);
			return [str, offset + n];
		})();

		[this.priority, offset] = (() => {
			const [val, n] = parseVarint(buf, offset);
			return [val, offset + n];
		})();

		[this.groupSequence, offset] = (() => {
			const [val, n] = parseVarint(buf, offset);
			return [val, offset + n];
		})();

		if (offset !== msgLen) {
			return new Error("message length mismatch");
		}

		return undefined;
	}
}
